### ISSUE TITLE
Base board evolution and secondary camera (CAM0)

### ISSUE THREAD

**Issue by bitbashing** [magkopian/raspberry-pi-compute-module-base-board#3]
Kudos @magkopian for creating this base board and for sharing it with us here.

The CM4 revision is now out (with the KiCad schematic entirely open sourced) but this is still of value to those tinkerers who have a few CM3/CM3+ boards laying around. Two questions for you:

In your real world experience using this base board, did you find it suitable for your needs or have you felt like you needed to make (or have you made since) any modifications to rev A.? If so, could you post your updates to this repo.

Secondly: the CM modules, contrasted with the mainstream A/B boards, offer hardware support for two cameras; I noticed you're only breaking out one of the cameras in Rev. A. I also see there's room on the board for the second connector, if one were inclined to reuse the same board dimensions. From your experience, any issues or considerations with adding another connector and traces for CAM0 to this base board design?

Thanks a bunch

**Comment by magkopian** [comment]
> The CM4 revision is now out (with the KiCad schematic entirely open sourced) but this is still of value to those tinkerers who have a few CM3/CM3+ boards laying around.

Yes, when the CM4 was released and I read that the Raspberry Pi Foundation also released KiCAD design files for the IO board I was super happy. That was exactly what I was hopping to happen when I released my own project.

> In your real world experience using this base board, did you find it suitable for your needs or have you felt like you needed to make (or have you made since) any modifications to rev A.? If so, could you post your updates to this repo.

No, I haven’t made any modifications and if I had I've already posted a new revision. Also, now that the CM4 is also out a new revision of this board is unlikely.

> From your experience, any issues or considerations with adding another connector and traces for CAM0 to this base board design?

The only issue, is that you will most likely have to drop the pin-to-pin compatibility of the 40-pin GPIO header with the regular Raspberry Pi boards. If this is not a feature that you need, which most likely isn’t unless you plan using HATs designed for the regular Raspberry Pi boards with your design, then there should be no problem.

To use the second camera you also need to modify the pin definitions in the device tree, but I'm sure you're already aware of that and also the official documentation is very comprehensive on the matter.

**Comment by bitbashing** [comment]
> No, I haven’t made any modifications and if I had I've already posted a new revision. Also, now that the CM4 is also out a new revision of this board is unlikely.

Understood. I was just making sure there weren't any in-flight changes before I fork this; thanks for your diligence.

> The only issue, is that you will most likely have to drop the pin-to-pin compatibility of the 40-pin GPIO header with the regular Raspberry Pi boards. If this is not a feature that you need, which most likely isn’t unless you plan using HATs designed for the regular Raspberry Pi boards with your design, then there should be no problem.

Good point. Indeed, I noticed that the 2nd I2C port is taken up by the 2nd camera if configured. I've made two GPIO pins from the original pool only available for CAM0 but will allow its SDA/SCL pins to accept other I2C devices on the 40-pin GPIO header, for the compatibility you've mentioned.

> To use the second camera you also need to modify the pin definitions in the device tree, but I'm sure you're already aware of that and also the official documentation is very comprehensive on the matter.

Thanks, yeah I was previously able to configure the blob using the original I/O board for both Raspberry Pi OS _and_ Ubuntu 20.

**Comment by bitbashing** [comment]
I've made changes to the forked schematic to bring in CAM0 and switch CAM1 and CAM0 pin assignments to match the Raspberry Pi [instructions on the foundation website](https://www.raspberrypi.org/documentation/hardware/computemodule/cmio-camera.md) (for the reason that it might be easier to follow and/or troubleshoot for folks who are just starting out with this board revision or simply desire compatibility between this board and the RPi I/O board).
The fork for the CAM0+CAM1 version is located at https://github.com/blazczak/raspberry-pi-compute-module-base-board/

Could you take a look @magkopian? The remaining step is putting down the actual tracks in the existing layout, but being an EE novice I'm a bit apprehensive with that step.

**Comment by magkopian** [comment]
Unless I'm missing something I don't see any issues with the schematic. As for the PCB layout my advice is to try and keep all the traces as short as possible and avoid using vias if you can help it. Also, make sure you have a solid ground plane going underneath the signals and use the built-in feature in KiCAD to route the differential signals as pairs. It's really not that hard as it sounds, as long as you stick to these rules you should be fine.

**Comment by bitbashing** [comment]
Thanks for the pointers. I have the first revision of the PCB layout for CAM0+CAM1.

I kept getting mapping errors when opening the project in KiCAD 5 and since I'm new at this I didn't want to risk propagating an incorrect auto remapping so I didn't proceed until I found a resolution - it took me a few tries but I realized that I should just open this pre-existing project in KiCAD 4.x to force the use of old library definitions.

Keep in mind I'm a much better software engineer than electrical engineer. 😃 I tried to keep the majority of your original layout intact to keep "damage" at a minimum, so as a result the CAM0 traces are not necessarily the shortest. I also didn't end up using the auto feature for differential signals, and I guess if they are of equal lengths then I may have a problem depending on the data transmit speed? One remapping for the USB connector was force-rescued by KiCAD 4.0.7.

@magkopian could you take a look at the modified pcb in the above repo and let me know your comments/suggestions?

**Comment by bitbashing** [comment]
KiCAD has some annoying UI bugs, at least on Mac: I had to redo large parts of track segments as it considered them connected for the purpose of ratsnest but the total length calculator was not "seeing" all segments in a particular track.

Added some finishing touches, including tuning the lengths of CAM0 differential pairs and revving the board code. I will create a PR to have this all reviewed and hopefully merged in.
